server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java

```diff
@@ -916,6 +919,47 @@ public ReplicationGroup getReplicationGroup() {
         return replicationGroup;
     }
 
+    /**
+     * Returns the pending replication actions for the shard.
+     *
+     * @return the pending replication actions
+     */
+    public PendingReplicationActions getPendingReplicationActions() {
+        assert primaryMode;
+        return pendingReplication;
+    }
+
+    private void updateReplicationGroupAndNotify(boolean replicasMightHaveChanged) {
+        ReplicationGroup newReplicationGroup = calculateReplicationGroup();
+        ReplicationGroup oldReplicationGroup = this.replicationGroup;
+        if (replicasMightHaveChanged) {
+            Set<String> oldReplicaNodeIds;
+            if (oldReplicationGroup == null) {
+                oldReplicaNodeIds = Collections.emptySet();
+            } else {
+                oldReplicaNodeIds = oldReplicationGroup.getReplicationTargets().stream()
+                    .map(ShardRouting::currentNodeId)
+                    .collect(Collectors.toSet());
+            }
+            for (ShardRouting replica : newReplicationGroup.getReplicationTargets()) {
+                if (oldReplicaNodeIds.contains(replica.currentNodeId()) == false) {
+                    pendingReplication.nodeJoinedReplicationGroup(replica.currentNodeId());
+                }
+            }
+        }
+        this.replicationGroup = newReplicationGroup;
+        if (oldReplicationGroup != null && replicasMightHaveChanged) {
+            Set<String> newReplicaNodeIds = newReplicationGroup.getReplicationTargets().stream()
+                .map(ShardRouting::currentNodeId)
+                .collect(Collectors.toSet());
+            for (ShardRouting replica : oldReplicationGroup.getReplicationTargets()) {
```

Comment:
maybe calculate the node sets for both old and new, and then just use `Sets.difference` to determine what changed? With the approach here, you're notifying the absence of a given node multiple times.

Reply:
I'm not sure I follow this comment. We need to check the new ones against previous and add new nodes. And then check previous against new ones and remove old nodes. Maybe you can update this comment based on the current version.